autogalaxy/analysis/analysis.py

```diff
@@ -57,8 +60,134 @@ def plane_via_instance_from(self, instance: af.ModelInstance) -> Plane:
         An instance of the Plane class that is used to then fit the dataset.
         """
         if hasattr(instance, "clumps"):
-            return Plane(galaxies=instance.galaxies + instance.clumps)
-        return Plane(galaxies=instance.galaxies)
+            return Plane(
+                galaxies=instance.galaxies + instance.clumps,
+                run_time_dict=run_time_dict,
+            )
+        return Plane(galaxies=instance.galaxies, run_time_dict=run_time_dict)
+
+    @property
+    def fit_func(self) -> Callable:
+        raise NotImplementedError
+
+    def profile_log_likelihood_function(
+        self, instance: af.ModelInstance, paths: Optional[af.DirectoryPaths] = None
+    ) -> Tuple[Dict, Dict]:
+        """
+        This function is optionally called throughout a model-fit to profile the log likelihood function.
+
+        All function calls inside the `log_likelihood_function` that are decorated with the `profile_func` are timed
+        with their times stored in a dictionary called the `run_time_dict`.
+
+        An `info_dict` is also created which stores information on aspects of the model and dataset that dictate
+        run times, so the profiled times can be interpreted with this context.
+
+        The results of this profiling are then output to hard-disk in the `preloads` folder of the model-fit results,
+        which they can be inspected to ensure run-times are as expected.
+
+        Parameters
+        ----------
+        instance
+            An instance of the model that is being fitted to the data by this analysis (whose parameters have been set
+            via a non-linear search).
+        paths
+            The PyAutoFit paths object which manages all paths, e.g. where the non-linear search outputs are stored,
+            visualization and the pickled objects used by the aggregator output by this function.
+
+        Returns
+        -------
+        Two dictionaries, the profiling dictionary and info dictionary, which contain the profiling times of the
+        `log_likelihood_function` and information on the model and dataset used to perform the profiling.
+        """
+        run_time_dict = {}
+        info_dict = {}
+
+        repeats = conf.instance["general"]["profiling"]["repeats"]
+        info_dict["repeats"] = repeats
+
+        fit = self.fit_func(instance=instance)
+        fit.figure_of_merit
+
+        start = time.time()
+
+        for i in range(repeats):
+            try:
+                fit = self.fit_func(instance=instance)
+                fit.figure_of_merit
+            except Exception:
+                logger.info(
+                    "Profiling failed. Returning without outputting information."
+                )
+                return
+
+        fit_time = (time.time() - start) / repeats
+
+        run_time_dict["fit_time"] = fit_time
+
+        fit = self.fit_func(instance=instance, run_time_dict=run_time_dict)
+        fit.figure_of_merit

```

Comment:
Does this do anything?

Reply:
Ensures numba functions are compiled before profiling begins, comment added